apps/omg/test/omg/state/persistence_test.exs

```diff
@@ -33,172 +34,205 @@ defmodule OMG.State.PersistenceTest do
+    {:ok, _, _} = OMG.State.exit_utxos([Utxo.position(1, 0, 0)])
 
-    state
+    # exit db_updates won't get persisted yet, but alice tries to spent it immediately
+    assert :utxo_not_found == exec(create_recovered([{1, 0, 0, alice}], @eth, [{alice, 20}]))
+
+    # retry above with empty in-memory utxoset
+    :ok = restart_state()
+
+    {:ok, _, _} = OMG.State.exit_utxos([Utxo.position(1, 0, 0)])
+    assert :utxo_not_found == exec(create_recovered([{1, 0, 0, alice}], @eth, [{alice, 20}]))
   end
 
-  defp persist_deposit(state, deposits, db_pid) do
-    {:ok, {_, db_updates}, state} = Core.deposit(deposits, state)
-    persist_common(state, db_updates, db_pid)
+  defp persist_deposit(deposits) do
+    {:ok, db_updates} =
+      deposits
+      |> make_deposits()
+      |> OMG.State.deposit()
+
+    :ok = OMG.DB.multi_update(db_updates)
   end
 
-  defp persist_form(state, db_pid) do
-    {:ok, {_, _, db_updates}, state} = Core.form_block(@interval, state)
-    persist_common(state, db_updates, db_pid)
+  defp persist_form(:ok), do: persist_form()
+
+  defp persist_form() do
+    OMG.State.form_block()
+
+    # because `form_block` is non-blocking operation we need to wait it finishes
+    # easiest to do this is to schedule another operation on `OMG.State` which is blocking
+    _ = OMG.State.utxo_exists?(Utxo.position(0, 0, 0))
+
+    :ok
   end
 
-  defp persist_exit_utxos(state, utxo_positions, db_pid) do
-    assert {:ok, {db_updates, _}, state} = utxo_positions |> Core.exit_utxos(state)
-    persist_common(state, db_updates, db_pid)
+  defp exec(:ok, tx), do: exec(tx)
+
+  defp exec(tx) do
+    case OMG.State.exec(tx, :no_fees_required) do
+      {:ok, _} -> :ok
+      {:error, reason} -> reason
+    end
   end
 
-  defp exec(state, tx) do
-    assert {:ok, _, state} = Core.exec(state, tx, :no_fees_required)
-    state
+  defp persist_exit_utxos(:ok, exit_infos), do: persist_exit_utxos(exit_infos)
+
+  defp persist_exit_utxos(exit_infos) do
+    {:ok, db_updates, _} = OMG.State.exit_utxos(exit_infos)
+
+    :ok = OMG.DB.multi_update(db_updates)
   end
 
-  defp persisted_standard_exitable_utxos(address, db_pid) do
-    {:ok, utxos_query_result} = OMG.DB.utxos(db_pid)
-    Core.standard_exitable_utxos(utxos_query_result, address.addr)
+  defp make_deposits(list) do
+    list
+    |> Enum.map(fn %{owner: owner, currency: currency, amount: amount, blknum: blknum} ->
+      %{
+        root_chain_txhash: <<blknum::256>>,
+        log_index: 0,
+        owner: owner.addr,
+        currency: currency,
+        amount: amount,
+        blknum: blknum
+      }
+    end)
   end
 
-  defp persist_common(state, db_updates, db_pid) do
-    assert :ok = OMG.DB.multi_update(db_updates, db_pid)
-    assert state == state_from(db_pid)
-    state
+  defp restart_state() do
+    GenServer.stop(OMG.State)
+    WaitFor.ok(fn -> if(GenServer.whereis(OMG.State), do: :ok) end)
```

Comment:
`stop/1,2` is a synchronous operation.
https://hexdocs.pm/elixir/GenServer.html#stop/3

Reply:
That's correct - `WaitFor` is here to assume the supervisor bring the process back

Reply:
Why the assert? 
`assert OMG.State |> GenServer.whereis() |> Kernel.is_pid()`

Reply:
To double check :smile: 
I'll drop the `assert` line